htlcswitch/htlcnotifier.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	log.Tracef("Notifying forward event type: %v from %v(%v) to %v(%v)",
+		event.HTLCEventType, event.IncomingChannel, event.HTLCInIndex,
+		event.OutgoingChannel, event.HTLCOutIndex)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send forwarding event: %v", err)
+	}
+}
+
+// NotifyLinkFailEvent notifies the HTLCNotifier that we have failed a HTLC on
+// one of our links.
+func (h *HTLCNotifier) NotifyLinkFailEvent(pkt *htlcPacket, hash lntypes.Hash,
+	eventType HTLCEventType, incomingFailed bool) {
+
+	// Sanity check that the packet's link failure is provided.
+	if pkt.linkFailure == nil {
+		log.Error("NotifyLinkFailEvent link failure should be non-nil")
+		return
+	}
+
+	event := LinkFailEvent{
+		HTLCKey:        newHTLCKey(pkt),
+		HTLCInfo:       newHTLCInfo(pkt, hash, eventType),
+		FailReason:     pkt.linkFailure,
+		Timestamp:      time.Now(),
+		IncomingFailed: true,
+	}
+
+	// If the link failure is a forwarding event which enriches the lnwire
+	// failure with further details, set the values on the event.
+	fwdErr, ok := pkt.linkFailure.(*ForwardingError)
+	if ok {
+		event.FailReason = fwdErr.FailureMessage
+		event.FailDetail = fwdErr.FailureDetail
+	}
+
+	log.Tracef("Notifying link failure event type: %v from %v(%v) to "+
+		"%v(%v), incoming link: %v, with reason: %v, details: %T",
+		event.HTLCEventType, event.IncomingChannel, event.HTLCInIndex,
+		event.OutgoingChannel, event.HTLCOutIndex, event.IncomingFailed,
+		event.FailReason, event.FailDetail)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send link fail event: %v", err)
+	}
+}
+
+// NotifyForwardingFailEvent notifies the HTLCNotifier that a HTLC we forwarded
+// has failed down the line.
+func (h *HTLCNotifier) NotifyForwardingFailEvent(pkt *htlcPacket) {
+	// If the packet has a non-zero outgoing channel ID, it is a forward. Sends
+	// from our node that fail have zero outgoing channels IDs because we are
+	// the final node. We do not forward receives to our node, so we cannot have
+	// forwarding failures for receives.
+	eventType := HTLCEventTypeSend
+	if pkt.outgoingChanID.ToUint64() == 0 {
+		eventType = HTLCEventTypeForward
+	}
+
+	event := ForwardingFailEvent{
+		HTLCKey: newHTLCKey(pkt),
+		// HTLC hash in not available for forwarding failures, it can be obtained
+		// by matching the failure with its corresponding forward event.
+		HTLCInfo:  newHTLCInfo(pkt, lntypes.Hash{}, eventType),
+		Timestamp: time.Now(),
+	}
+
+	log.Tracef("Notifying forwarding failure event type: %v from %v(%v) "+
+		"to %v(%v)", event.HTLCEventType, event.IncomingChannel,
+		event.HTLCInIndex, event.OutgoingChannel, event.HTLCOutIndex)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send forwarding fail event: %v", err)
+	}
+}
+
+// NotifySettleEvent notifies the HTLCNotifier that a HTLC that we committed to
+// as part of a forward or a receive to our node has been settled.
```

Comment:
To clarify, it is us who settled the htlc right? Not our peer settling our htlc?